crates/oxc_span/src/ident.rs

```diff
@@ -0,0 +1,384 @@
+    }
+}
+
+impl<'a> From<&'a str> for Ident<'a> {
+    #[expect(clippy::inline_always)]
+    #[inline(always)] // Because this is a no-op
+    fn from(s: &'a str) -> Self {
+        Self::new(s)
+    }
+}
+
+impl<'a> From<Ident<'a>> for Atom<'a> {
+    #[inline]
+    fn from(val: Ident<'a>) -> Self {
+        val.into_atom()
+    }
+}
+
+impl<'a> From<Atom<'a>> for Ident<'a> {
+    #[inline]
+    fn from(atom: Atom<'a>) -> Self {
+        Self::new(atom.as_str())
+    }
+}
+
+impl<'a> From<Ident<'a>> for CompactStr {
+    #[inline]
+    fn from(ident: Ident<'a>) -> Self {
+        ident.into_compact_str()
+    }
+}
+
+impl<'a> From<Ident<'a>> for Cow<'a, str> {
+    #[expect(clippy::inline_always)]
+    #[inline(always)] // Because this is a no-op
+    fn from(ident: Ident<'a>) -> Self {
+        Cow::Borrowed(ident.as_str())
+    }
+}
+
+impl<'alloc> FromIn<'alloc, &Ident<'alloc>> for Ident<'alloc> {
+    #[expect(clippy::inline_always)]
+    #[inline(always)] // Because this is a no-op
+    fn from_in(s: &Ident<'alloc>, _: &'alloc Allocator) -> Self {
+        *s
+    }
+}
+
+impl<'alloc> FromIn<'alloc, &str> for Ident<'alloc> {
+    #[inline]
+    fn from_in(s: &str, allocator: &'alloc Allocator) -> Self {
+        Self::from(allocator.alloc_str(s))
+    }
+}
+
+impl<'alloc> FromIn<'alloc, String> for Ident<'alloc> {
+    #[inline]
+    fn from_in(s: String, allocator: &'alloc Allocator) -> Self {
+        Self::from(allocator.alloc_str(&s))
+    }
+}
+
+impl<'alloc> FromIn<'alloc, &String> for Ident<'alloc> {
+    #[inline]
+    fn from_in(s: &String, allocator: &'alloc Allocator) -> Self {
+        Self::from(allocator.alloc_str(s))
+    }
+}
+
+// TODO: Add rest of `FromIn` impls for `Ident`
+
+impl<'new_alloc> CloneIn<'new_alloc> for Ident<'_> {
+    type Cloned = Ident<'new_alloc>;
+
+    #[inline]
+    fn clone_in(&self, allocator: &'new_alloc Allocator) -> Self::Cloned {
+        Ident::from_in(self.as_str(), allocator)
+    }
+}
+
+impl<'a> Dummy<'a> for Ident<'a> {
+    /// Create a dummy [`Ident`].
+    #[expect(clippy::inline_always)]
+    #[inline(always)]
+    fn dummy(_allocator: &'a Allocator) -> Self {
+        Ident::new("x")
+    }
+}
+
+// Only use this `Hash` impl with `IdentHashMap`
+impl Hash for Ident<'_> {
+    fn hash<H: Hasher>(&self, hasher: &mut H) {
+        // Hash is stored in top 32 bits of `len_and_hash`.
+        // `hashbrown` only uses top 7 bits and bottom N bits of hash,
+        // where N is the exponent of number of buckets in the hash table.
+        // So just rotate by 25 bits, so 32-bit hash occupies top 7 bits
+        // and bottom 25 bits. This gives good entropy as long as HashMap
+        // contains less than 58 million entries ((1 << 25) * 7 / 8).
+        // The bits in the middle have low entropy, but that doesn't matter
+        // because `hashbrown` will ignore them anyway.
+        let hash = self.len_and_hash.rotate_left(32 - 7);
+        hasher.write_u64(hash);
+    }
+}
```

Comment:
All these small methods should have `#[inline]` on them. Probably compiler will inline them anyway, but just to make sure.